library/core/src/num/mod.rs

```diff
@@ -1581,3 +1585,143 @@ from_str_radix_size_impl! { i16 isize, u16 usize }
 from_str_radix_size_impl! { i32 isize, u32 usize }
 #[cfg(target_pointer_width = "64")]
 from_str_radix_size_impl! { i64 isize, u64 usize }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+mod verify {
+    use super::*;
+
+    // Verify `unchecked_{add, sub, mul}`
+    // However, `unchecked_mul` harnesses have bad performance, so
+    // recommend to use generate_unchecked_mul_harness! to set input limits
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>();
+                let num2: $type = kani::any::<$type>();
+
+                unsafe {
+                    num1.$method(num2);
+                }
+            }
+        }
+    }
+
+    // Improve unchecked_mul performance for {32, 64, 128}-bit integer types
+    // by adding upper and lower limits for inputs
+    macro_rules! generate_unchecked_mul_harness {
+        ($type:ty, $method:ident, $harness_name:ident, $min:expr, $max:expr) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>();
+                let num2: $type = kani::any::<$type>();
+
+                kani::assume(num1 >= $min && num1 <= $max);
+                kani::assume(num2 >= $min && num2 <= $max);
+    
+                unsafe {
+                    num1.$method(num2);
+                }
+            }
```

Comment:
Hi @celinval @feliperodri  , For the multiplication harnesses (128 bit), the proofs are taking a long time and not fully completing. Based on my research, I feel this is because the state space for checking 128 bit integers is extremely large, i.e. there should be approx 2 ( 2 ^ 128 ) = 2 ^ 256 number of combinations that are valid for verification, which is practically not possible. Hence I have limited the ranges to search within. Do you think this is a feasible approach, or am I missing something?

Note : i32, u32, i64 and u64 could still be run. But it would at least take a couple hours to finish. Is that acceptable? Or do we have any performance constraints to adhere to. Do let me know.

Thank you!

Reply:
Hi @rajathkotyal thanks for the question. I think adding a few intervals that cover different scenarios is a good compromise, such as using large numbers, small numbers, negatives, positives, inputs that would overflow.

Ideally, each harness should run under 2min. Let me know if that is not enough. Thanks!

Reply:
Thank you for the clarification, @celinval!

I have added the intervals for lower bounds, upper bounds, and edge checking. Please review once. Also, each check wrt multiplication runs well within 2 mins now.